core/src/main/scala/kafka/log/ProducerIdMapping.scala

```diff
@@ -54,48 +56,106 @@ private[log] class ProducerAppendInfo(val pid: Long, initialEntry: ProducerIdEnt
+        s"incomingBatch.lastSeq): ($firstSeq, $lastSeq), (lastEntry.firstSeq, lastEntry.lastSeq): " +
+        s"(${this.firstSeq}, ${this.lastSeq}).")
     } else if (firstSeq != this.lastSeq + 1L) {
-      throw new OutOfOrderSequenceException(s"Invalid sequence number: $pid (pid), $firstSeq " +
-        s"(seq. number), ${this.lastSeq} (expected seq. number)")
+      throw new OutOfOrderSequenceException(s"Out of order sequence number: $pid (pid), $firstSeq " +
+        s"(incoming seq. number), ${this.lastSeq} (current end sequence number)")
     }
   }
 
   def assignLastOffsetAndTimestamp(lastOffset: Long, lastTimestamp: Long): Unit = {
+    if (startedTxns.nonEmpty) {
+      startedTxns.clear()
+      val firstOffset = lastOffset - (lastSeq - firstSeq)
+      startedTxns += OngoingTxn(pid, firstOffset)
+      if (currentTxnFirstOffset == 0)
+        currentTxnFirstOffset = firstOffset
+    }
+
     this.lastOffset = lastOffset
     this.maxTimestamp = lastTimestamp
   }
 
-  private def append(epoch: Short, firstSeq: Int, lastSeq: Int, lastTimestamp: Long, lastOffset: Long) {
+  def append(epoch: Short, firstSeq: Int, lastSeq: Int, lastTimestamp: Long, lastOffset: Long, isTransactional: Boolean) {
     validateAppend(epoch, firstSeq, lastSeq)
     this.epoch = epoch
     this.firstSeq = firstSeq
     this.lastSeq = lastSeq
     this.maxTimestamp = lastTimestamp
     this.lastOffset = lastOffset
+
+    if (currentTxnFirstOffset >= 0 && !isTransactional)
+      // FIXME: Do we need a new error code here?
+      throw new InvalidRecordException(s"Expected transactional write from producer $pid")
+
+    if (isTransactional && currentTxnFirstOffset < 0) {
+      val firstOffset = lastOffset - (lastSeq - firstSeq)
+      currentTxnFirstOffset = firstOffset
+      startedTxns += OngoingTxn(pid, firstOffset)
+    }
   }
 
   def append(batch: RecordBatch): Unit =
-    append(batch.producerEpoch, batch.baseSequence, batch.lastSequence, batch.maxTimestamp, batch.lastOffset)
+    append(batch.producerEpoch, batch.baseSequence, batch.lastSequence, batch.maxTimestamp, batch.lastOffset,
+      batch.isTransactional)
+
+  def appendControl(controlRecord: ControlRecord): CompletedTxn = {
+    if (this.epoch > controlRecord.epoch)
+      throw new ProducerFencedException(s"Invalid epoch (zombie writer): ${controlRecord.epoch} (request epoch), ${this.epoch}")
+
+    controlRecord.controlType match {
+      case ControlRecordType.ABORT | ControlRecordType.COMMIT =>
+        val firstOffset = if (currentTxnFirstOffset >= 0)
+          currentTxnFirstOffset
+        else
+          controlRecord.offset
 
-  def append(entry: ProducerIdEntry): Unit =
-    append(entry.epoch, entry.firstSeq, entry.lastSeq, entry.timestamp, entry.lastOffset)
+        currentTxnFirstOffset = -1
+        maxTimestamp = controlRecord.timestamp
+        CompletedTxn(pid, firstOffset, lastOffset, controlRecord.controlType == ControlRecordType.ABORT)
```

Comment:
I've fixed this. We always use the offset of the control record as the last offset.